core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Wraps a {@code WatchService} and allows paths to be registered.
+ */
+final class PropertiesFileWatcherRegistry implements AutoCloseable {
+    private static final Logger logger = LoggerFactory.getLogger(PropertiesFileWatcherRegistry.class);
+
+    @Nullable
+    private CompletableFuture<Void> future;
+    private final Map<Path, PropertiesFileWatcherContext> ctxRegistry = new ConcurrentHashMap<>();
+    private final WatchService watchService;
+    private final ExecutorService executor =
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setDaemon(true).build());
```

Comment:
- Could you set the thread name?
- Could we make the executor is started lazily and shut down when there are no watches anymore, so that this class is more friendly to multi-classloader environment? For example, someone might be using Armeria as a part of a Java EE webapp and want Armeria to clean up everything correctly when the web application is reloaded.

Reply:
I was kind of thinking as long as a submitted `thread` (or `future`) isn't running, there should be minimal overhead from just instantiating `ExecutorService`. (The `future` is currently stopped when no paths are registered). Let me know if I am thinking about this wrong 🤔